cmd/orderimports/orderimports.go

```diff
@@ -0,0 +1,207 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// verify that all the imports have our preferred order.
+// https://github.com/kubernetes/kubeadm/issues/2515
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+var (
+	includePath = flag.String("include-path", "", "only files with paths matching this regex are touched")
+	ignoreFile  = flag.String("ignore-file", "zz_generated", "files matching this regex are ignored")
+)
+
+type analyzer struct {
+	fset       *token.FileSet // positions are relative to fset
+	failed     bool
+	ignoreFile *regexp.Regexp
+}
+
+func newAnalyzer() *analyzer {
+	ignoreFileRegexp, err := regexp.Compile(*ignoreFile)
+	if err != nil {
+		log.Fatalf("Error compiling ignore regex: %v", err)
+	}
+
+	a := &analyzer{
+		fset:       token.NewFileSet(),
+		ignoreFile: ignoreFileRegexp,
+	}
+
+	return a
+}
+
+// collect extracts test metadata from a file.
+func (a *analyzer) collect(dir string) {
+	// create the AST by parsing src.
+	fs, err := parser.ParseDir(a.fset, dir, nil, parser.AllErrors|parser.ParseComments)
+
+	if err != nil {
+		fmt.Println(err)
+		a.failed = true
+		return
+	}
+
+	for _, p := range fs {
+		files := a.filterFiles(p.Files)
+		for _, file := range files {
+			pathToFile := a.fset.File(file.Pos()).Name()
+
+			if len(file.Imports) <= 1 {
+				continue
+			}
+			var originalImports, stdlibImports, localImports, k8sImports, externalImports []string
+
+			for i, imp := range file.Imports {
+				importPath := strings.Replace(imp.Path.Value, "\"", "", -1)
+				parts := strings.Split(importPath, "/")
+
+				// if the original imports have blank line, need to add a blank line for originalImports too
+				if i != 0 && a.lineAt(imp.Pos()) > 1+a.lineAt(file.Imports[i-1].End()) {
+					originalImports = append(originalImports, "")
+				}
+				originalImports = append(originalImports, importPath)
+
+				if !strings.Contains(parts[0], ".") {
+					// standard library imports are first
+					stdlibImports = append(stdlibImports, importPath)
+				} else if strings.HasPrefix(importPath, "k8s.io/kubernetes") {
+					// local imports are second
+					localImports = append(localImports, importPath)
+				} else if strings.Contains(parts[0], "k8s.io") {
+					// other *.k8s.io imports are third
+					k8sImports = append(k8sImports, importPath)
+				} else {
+					// external repositories are fourth
+					externalImports = append(externalImports, importPath)
+				}
```

Comment:
this doesn't match the typical project order:

```
stdlib imports

imports outside k8s.io/kubernetes

imports within k8s.io/kubernetes
```

was that intentional?

Reply:
i was under the impression that k/k doesn't have a typical import order, which is confirmed by looking at some files under /pkg.

i proposed this order here:
https://github.com/kubernetes/kubeadm/issues/2515
no strong preference on my side in terms of the ordering, as long there is separation.

> imports outside k8s.io/kubernetes

IMO, there should be two distinct sections: *.k8s.io imports, third party imports

Reply:
> IMO, there should be two distinct sections: *.k8s.io imports, third party imports

stdlib / non-k8s / k8s groupings make sense to me. it looks like some files split out of repo k8s stuff (like k8s.io/utils and k8s.io/klog) out and some don't.

Reply:
ok, @xinydev the new order should be:

```
stdlib

imports outside of *.k8s.io

*.k8s.io (not local repository)

local repository
```

i guess we also need to re-order /cmd/kubeadm again in the second commit.